File: C++/Singleton/Singleton.cpp
#include <iostream>
using namespace std;

class Singleton
{
    public:
    static Singleton & GetInstance()
    {
        cout << "Singleton GetInstance" << endl;
        if (m_pSingelton == nullptr)
        {
            m_pSingelton = new Singleton();
        }
        return *m_pSingelton;
    }
    static void DeleteInstance()
    {
        cout << "Singleton DeleteInstance" << endl;
        delete m_pSingelton;
        m_pSingelton = nullptr;
    }
    ~Singleton()
    {
        cout << "Singleton : Destructor" << endl;
    }

    // deleting copy constructor
    Singleton(const Singleton& obj)= delete;

    //Singletons should not be assignable.
    void operator=(const Singleton &) = delete;
    
    private:
    // private constructor so that no one can initialize the object outside directly
    Singleton()
    {
        cout << "Singelton : Constructor" << endl;
    }

    static Singleton * m_pSingelton;
    protected:
};

Singleton * Singleton::m_pSingelton = nullptr;

int main()
{
    Singleton &pSingelton = Singleton::GetInstance();
    Singleton::DeleteInstance();
    
    return 0;
}

File: Java/Singleton/Singleton.java
public class Singleton
{
    static private Singleton m_Singleton;
    private int _data;
    static public Singleton GetInstance()
    {
        System.out.println("Singleton GetInstance");
        if (m_Singleton == null)
        {
            m_Singleton = new Singleton();
        }
        return m_Singleton;
    }

    void setdata(int data)
    {
        _data = data;
    }

    int getdata()
    {
        return _data;
    }

    private Singleton()
    {
        System.out.println("Singleton Constructor");
    }
    public static void main(String[] args)
    {
        Singleton singleton1 = Singleton.GetInstance();
        singleton1.setdata(5);
        System.out.println("Set data: 5");

        Singleton singleton2 = Singleton.GetInstance();
        System.out.println("Get data: " +singleton2.getdata());
        return;
    }
}